### PR TITLE
[TEST] Allow modules with non-leading batch axis 

### PR DESCRIPTION
**Pull request by tjs111**
In preparation of support for RNN, GRU, and LSTM and the possibility of a different position of the batch axis.
This corresponds to issue fKunstner/backpack-discuss#100

### CODE REVIEW

**Comment by f-dangel** [comment]
Waiting for the same functionality being introduced to the `extension`s test suite.

File: test/core/derivatives/implementation/autograd.py [hunk 1]
```diff
@@ -40,17 +40,35 @@ def weight_jac_t_mat_prod(self, mat, sum_batch):
     def bias_jac_t_mat_prod(self, mat, sum_batch):
         return self.param_jac_t_mat_prod("bias", mat, sum_batch)
 
-    def param_jac_t_vec_prod(self, name, vec, sum_batch):
+    def bias_ih_l0_jac_t_mat_prod(self, mat, sum_batch):
+        return self.param_jac_t_mat_prod("bias_ih_l0", mat, sum_batch, axis_batch=1)
+
+    def param_jac_t_vec_prod(self, name, vec, sum_batch, axis_batch=0):
+        """
+        Compute the product of jac_t and the given vector.
```
Comment (f-dangel) [inline test/core/derivatives/implementation/autograd.py:48]: Remove the blank line, `"""Compute ...`

File: test/core/derivatives/implementation/autograd.py [hunk 2]
```diff
@@ -40,17 +40,35 @@ def weight_jac_t_mat_prod(self, mat, sum_batch):
...
+
+        Args:
+            name (str): name of parameter for derivative
+            vec (torch.Tensor): vectors which to multiply
+            sum_batch (boolean): whether to sum along batch axis
+            axis_batch (int, optional): index of batch axis. Defaults to 0.
+        Returns:
```
Comment (f-dangel) [inline test/core/derivatives/implementation/autograd.py:55]: Blank line before `Returns:`

File: test/core/derivatives/implementation/autograd.py [hunk 3]
```diff
@@ -59,12 +77,25 @@ def param_jac_t_vec_prod(self, name, vec, sum_batch):
...
+        Compute the product of jac_t and the given matrix.
+
+        Args:
+            name (str): name of parameter for derivative
+            mat (torch.Tensor): matrix which to multiply
+            sum_batch (boolean): whether to sum along batch axis
+            axis_batch (int, optional): index of batch axis. This is counted
+                without the first axis. Defaults to 0.
+        Returns:
+            torch.Tensor: product of jac_t and mat
```
Comment (f-dangel) [inline test/core/derivatives/implementation/autograd.py:91]: Please modify this docstring as described above, too

File: test/core/derivatives/implementation/autograd.py [hunk 4]
```diff
@@ -40,17 +40,35 @@ def weight_jac_t_mat_prod(self, mat, sum_batch):
...
         if sum_batch:
             return transposed_jacobian_vector_product(output, param, vec)[0]
         else:
-            N = input.shape[0]
-
-            sample_outputs = [output[n] for n in range(N)]
-            sample_vecs = [vec[n] for n in range(N)]
+            sample_outputs = [
+                tensor.squeeze(dim=axis_batch)
+                for tensor in torch.split(output, 1, dim=axis_batch)
```
Comment (f-dangel) [inline test/core/derivatives/implementation/autograd.py:66]: Not sure if it works, but `output.split(1, dim=axis_batch)` would be shorter

File: test/core/derivatives/implementation/autograd.py [hunk 5]
```diff
@@ -40,17 +40,35 @@ def weight_jac_t_mat_prod(self, mat, sum_batch):
...
+            ]
+            sample_vecs = [
+                tensor.squeeze(dim=axis_batch)
+                for tensor in torch.split(vec, 1, dim=axis_batch)
```
Comment (f-dangel) [inline test/core/derivatives/implementation/autograd.py:70]: Not sure if it works, but `vec.split(1, dim=axis_batch)` would be shorter

File: test/extensions/implementation/autograd.py [hunk 6]
```diff
@@ -11,7 +11,12 @@ class AutogradExtensions(ExtensionsImplementation):
     """Extension implementations with autograd."""
 
     def batch_grad(self):
-        N = self.problem.input.shape[0]
+        """Calculates the individual gradients.
```
Comment (f-dangel) [inline test/extensions/implementation/autograd.py:14]: scaled individual gradients

File: test/extensions/implementation/autograd.py [hunk 7]
```diff
@@ -11,7 +11,12 @@ class AutogradExtensions(ExtensionsImplementation):
...
+
+        Returns:
+            list[torch.nn.Tensor]: batch_grads
```
Comment (f-dangel) [inline test/extensions/implementation/autograd.py:17]: Scaled individual gradients comptuted by BackPACK's `BatchGrad` extension

File: test/extensions/problem.py [hunk 8]
```diff
@@ -9,6 +9,14 @@
 
 
 def make_test_problems(settings):
+    """Creates test problems from settings.
+
+    Args:
+        settings (list): raw settings of the problems
+
+    Returns:
+        list
```
Comment (f-dangel) [inline test/extensions/problem.py:18]: Can you add a description of the list items, i.e. what they contain?

File: test/extensions/problem.py [hunk 9]
```diff
@@ -108,27 +133,56 @@ def make_id(self):
         )
 
     def forward_pass(self, sample_idx=None):
-        """Do a forward pass. Return input, output, and parameters."""
+        """Do a forward pass. Return input, output, and parameters.
+
+        Args:
+            sample_idx (int): index of the sample to select
```
Comment (f-dangel) [inline test/extensions/problem.py:139]: Add optional and describe what happens in default scenario

File: test/extensions/problem.py [hunk 10]
```diff
@@ -108,27 +133,56 @@ def make_id(self):
...
+                input, output, loss, each with batch axis first
+        """
         if sample_idx is None:
             input = self.input.clone().detach()
             target = self.target.clone().detach()
         else:
-            input = self.input.clone()[sample_idx, :].unsqueeze(0).detach()
             target = self.target.clone()[sample_idx].unsqueeze(0).detach()
+            input = torch.split(
+                self.input, split_size_or_sections=1, dim=self.axis_batch
```
Comment (f-dangel) [inline test/extensions/problem.py:151]: - `self.input.split(...)` is shorter (not sure if it works)
- make call consistent to calls in core test suite by replacing `split_size_or_sections=1` by `1`

File: test/extensions/problem.py [hunk 11]
```diff
@@ -108,27 +133,56 @@ def make_id(self):
...
         self.model = extend(self.model)
         self.loss_function = extend(self.loss_function)
 
     def get_reduction_factor(self, loss, unreduced_loss):
-        """Return the factor used to reduce the individual losses."""
+        """Return the factor used to reduce the individual losses.
+
+        Args:
+            loss: the loss after reduction
+            unreduced_loss: the raw loss before reduction
```
Comment (f-dangel) [inline test/extensions/problem.py:178]: Add type specification `torch.Tensor`

File: test/extensions/problem.py [hunk 12]
```diff
@@ -108,27 +133,56 @@ def make_id(self):
...
+
+        Returns:
+            long: factor
```
Comment (f-dangel) [inline test/extensions/problem.py:181]: Not sure, but should be a scalar `torch.Tensor`

File: test/core/derivatives/implementation/autograd.py [hunk 13]
```diff
@@ -40,17 +40,35 @@ def weight_jac_t_mat_prod(self, mat, sum_batch):
...
-            N = input.shape[0]
-
-            sample_outputs = [output[n] for n in range(N)]
-            sample_vecs = [vec[n] for n in range(N)]
+            sample_outputs = [
+                tensor.squeeze(dim=axis_batch)
+                for tensor in output.split(1, dim=axis_batch)
+            ]
+            sample_vecs = [
+                tensor.squeeze(dim=axis_batch)
```
Comment (f-dangel) [inline test/core/derivatives/implementation/autograd.py:69]: I know the old implementation squeezes out the batch dimension. However, I think it's safer to explicitly keep it and therefore effectively work with `batch_size=1`, as it avoids different behavior of layers on unbatched inputs. Can you remove the `squeeze`? The tests should still pass.

File: test/extensions/problem.py [hunk 14]
```diff
@@ -9,6 +9,14 @@
 
 
 def make_test_problems(settings):
+    """Creates test problems from settings.
+
+    Args:
+        settings (list[dict]): raw settings of the problems
+
+    Returns:
+        list[ExtensionTestProblem]
```
Comment (f-dangel) [inline test/extensions/problem.py:18]: Same as above

File: test/extensions/problem.py [hunk 15]
```diff
@@ -108,27 +133,58 @@ def make_id(self):
...
+
+        The forward pass is performed on the selected index.
+        If the index is None, then the forward pass is calculated for the whole batch.
+
+        Args:
+            sample_idx (int, optional): Index of the sample to select.
+                Defaults to None.
+
+        Returns:
+            tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
```
Comment (f-dangel) [inline test/extensions/problem.py:146]: Same as above

File: test/core/derivatives/implementation/autograd.py (hunk 13 above)
Comment (tjs111) [reply test/core/derivatives/implementation/autograd.py:69]: Leads to a failed test.
When changing the line to `tensor` instead of `tensor.squeeze(dim=axis_batch)` the first test that fails in `make test-light` is `test/core/derivatives/derivatives_test.py::test_weight_jac_t_mat_prod[conv-example-dev=cpu-in=(3,2,7,7)-Conv2d(2,3,kernel_size=(2,2),stride=(2,2),padding=(1,1),dilation=(2,2),bias=False)-save_memory=True-sum_batch=False]` with error RuntimeError shape mismatch.

Comment (f-dangel) [reply test/core/derivatives/implementation/autograd.py:69]: Oh, so then there is a more general problem. Let's discuss in person.

File: test/extensions/implementation/autograd.py [hunk 16]
```diff
@@ -11,7 +11,12 @@ class AutogradExtensions(ExtensionsImplementation):
     """Extension implementations with autograd."""
 
     def batch_grad(self):
-        N = self.problem.input.shape[0]
+        """Scaled individual gradients computed by BackPACK's BatchGrad extension.
+
+        Returns:
+            list[torch.Tensor]: batch_grads
```
[the remark this answers is not in the document]
Comment (tjs111) [reply test/extensions/implementation/autograd.py:17]: Yes, I think it is the correct format.
My PyCharm compiler set to docstring format "Google" exclusively accepts it this way.
[Stackoverflow](https://stackoverflow.com/questions/41055326/how-to-denote-return-type-tuple-in-google-style-pydoc-for-pycharm) also says so.

File: test/core/derivatives/implementation/autograd.py (hunk 13 above)
Comment (tjs111) [reply test/core/derivatives/implementation/autograd.py]: Note: Changing for sample_vecs and sample_outputs works.
